### ISSUE TITLE
Problems with reverse proxying mstream  through nginx

### ISSUE THREAD

**Issue by Nixbligger** [IrosTheBeggar/mStream#200]
Hello there,

I am trying to setup mstream on a similar setup as in #64 but when I try to access my server subdomain https://server/mstream the request gets instantly redirected to https://server/access-denied what gives a 404 from nginx. If I proxy_pass this page via nginx config, I get a blank json page stating:
error | "Access Denied"  (btw. the same as if I try to open https://darncoyotes.mstream.io/)

Expected behavior: get a login page of some sort presented on accessing …/mstream

What am I doing wrong?

I read something about mstream not being able to get served via (apache) reverse proxy / dir mappings in #170, but I am not sure if this applies to my setup with nginx. But to my understanding I do not have a real subdomain configured for mstream (I cannot as I am not the owner of the domain). Would you say the #170 restriction applies to my situation?

Thank you for reading!
Cheers, 
Nix

**Comment by Nixbligger** [comment]
Additional info:

nginx block

>        location /mstream {
>                 proxy_pass http://127.0.0.1:3000;
>                 proxy_http_version 1.1;
>                 proxy_set_header Host $host;
>                 proxy_set_header Upgrade $http_upgrade;
>                 proxy_set_header Connection "upgrade";
>         }

```
uname -a
Linux … 4.14.27-v7+ #1100 SMP Fri Mar 16 13:51:48 GMT 2018 armv7l GNU/Linux
```

```
docker -v
Docker version 18.09.0, build 4d60db4
```

```
docker logs --tail 20 mstream
…
2019-07-31T11:47:42.412Z info: Access mStream locally: http://localhost:3000
2019-07-31T11:47:42.426Z info: Try the WinAmp Demo: http://localhost:3000/winamp
2019-07-31T11:47:57.487Z info: File scan started on /music
```
```
curl http://localhost:3000
<!doctype html>
<html lang="en">

<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />

  <meta content="yes" name="mobile-web-app-capable">
  <meta content="yes" name="apple-mobile-web-app-capable">
  <meta content="black" name="apple-mobile-web-app-status-bar-style">
…
```

**Comment by IrosTheBeggar** [comment]
Hey Nix, thanks for checking out mStream.

mStream was never built for reverse proxy by directory mappings.  The app uses absolute URLS, which is why you see these problems. I built mStream with the idea users would be reverse proxying it to a sub-domain.

One user recently posted a pull request changes the app to use relative urls.  I have done some testing but not enough to merge it yet.  Give this a try and let me know if it works for use case. If it does, I'll make this a high priority issue (currently all my time is going into a backup/federation project).   https://github.com/IrosTheBeggar/mStream/pull/190

Also https://darncoyotes.mstream.io/ has been offline for a while.  The new domain is https://demo.mstream.io/.  Did I leave an old link on the docs or website?

**Comment by IrosTheBeggar** [comment]
There's a new branch called `relative-urls` that has the merged changes

**Comment by Nixbligger** [comment]
Hey Paul,

thanks for the quick and informative reply. No, I think I found the link in #64 . The newer link works flawlessly. It  convinced me to use mstream as a subsonic replacement in the first place.
I will check out the branch asap although I am not quite sure how to actually do it. I am not a programmer myself, and not so familiar with the git environment yet.
But I suppose I should go for a local instead of a docker install and pull this branch and try to make it work, right?

I will check back once I tried it.

Cheerio,
Nix

**Comment by IrosTheBeggar** [comment]
Git can appear daunting, so here's the basic for checking out a branch

```shell
# Clone mStream
git clone https://github.com/IrosTheBeggar/mStream.git

# Go into mstream directory
cd mstream

# Switch branch
git checkout relative-urls

# Switch back to the main branch
git checkout master
```

After that you need to go into the directory and install and run mstream

```shell
npm install

mstream -m /path/to/your/music
```

I'm writing that from memory, so let me know if you have any problems

**Comment by Nixbligger** [comment]
Hello again,

I just checked out the steps you wrote above, and the git stuff went smooth.
Following the respective chapter in the README.md I added `sudo npm link` to set the respective links. I hope that wasn't an error.
Unfortunately, it does return an error when I run `mstream -m /path/to/your/music`:

> …/mStream/modules/config/configure-commander.js:87
>   let program3 = {
>   ^^^
> 
> SyntaxError: Block-scoped declarations (let, const, function, class) not yet supported outside strict mode
>     at exports.runInThisContext (vm.js:53:16)
>     at Module._compile (module.js:373:25)
>     at Object.Module._extensions..js (module.js:416:10)
>     at Module.load (module.js:343:32)
>     at Function.Module._load (module.js:300:12)
>     at Module.require (module.js:353:17)
>     at require (internal/module.js:12:17)
>     at Object.<anonymous> (/home/larse/Temp/mStream/cli-boot-wrapper.js:11:15)
>     at Module._compile (module.js:409:26)
>     at Object.Module._extensions..js (module.js:416:10)

Any Ideas how to proceed from here? Thanks for the support!
Best,
Nix

**Comment by IrosTheBeggar** [comment]
That's an error I haven't seen in a while. Check your node version with `node -v`.  You will probably need to upgrade to node v10

**Comment by Nixbligger** [comment]
What, v10?
`node -v` on my Raspbian stretch  gives:

> v4.8.2

My goodness 8-) Dunno where it's from.
I suppose I'll have to look into this…

**Comment by Nixbligger** [comment]
Hello again,

I upgraded my nodejs installation to 12.x and mstream seems to start up well. The server is locally available on the specified port. But so far I haven't been able to reverse proxy it through my nginx instance. 
A otherwise entirely blanc page is providing the following reply on access:
> Cannot GET /mstream

Any ideas?

**Comment by Nixbligger** [comment]
Also, after a long while hanging on 2250 db entries, the scan for media ends up with the following error:

> <--- Last few GCs --->
> cr[26232:0x3c97ab0]  5262296 ms: Mark-sweep 297.2 (299.4) -> 297.1 (299.4) MB, 1382.6 / 0.0 ms  (+ 158.1 ms in 64 steps since start of marking, biggest step 16.7 ms, walltime since start of marking 1635 ms) (average mu = 0.117, current mu = 0.058) finalize [26232:0x3c97ab0]  5263815 ms: Mark-sweep 297.4 (299.4) -> 297.1 (299.4) MB, 1187.5 / 0.0 ms  (+ 145.6 ms in 85 steps since start of marking, biggest step 11.5 ms, walltime since start of marking 1518 ms) (average mu = 0.120, current mu = 0.123) allocatio
> 
> 2019-08-17T16:01:38.702Z info: File scan message: 
> <--- JS stacktrace --->
> 
> ==== JS stack trace =========================================
> 
>     0: ExitFrame [pc: 0xe1c140]
> Security context: 0x24ecdc19 <JSObject>
>     1: /* anonymous */(aka /* anonymous */) [0x293fa5c5] […/mStream/node_modules/strtok3/lib/FileTokenizer.js:~2] [pc=0x2bb998ac](this=0x35ac0289 <undefined>,0x4cc8be99 <AbstractTokenizer map = 0x2f9c626d>,0x35ac0289 <undefined>,0x35ac0289 <undefined>,0x4c1aec0d <JSFunction (sfi = 0x4f6979cd)>)
>     2: peekBuffer [0x293fa9cd] […/mStream/no...
> 
> 
> 2019-08-17T16:01:38.706Z error: File scan error: FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
> 
> 2019-08-17T16:01:38.834Z error: File scan error: 
> 
> 2019-08-17T16:01:38.852Z error: File scan error: Writing Node.js report to file: report.20190817.180138.26232.0.001.json
> 2019-08-17T16:01:38.938Z error: File scan error: 
> 
> 2019-08-17T16:01:38.940Z error: File scan error: Node.js report completed


Any ideas, Paul?

Cheers, Nix

**Comment by AlphaJack** [comment]
> But so far I haven't been able to reverse proxy it through my nginx instance.
> A otherwise entirely blanc page is providing the following reply on access:
> > Cannot GET /mstream
> Any ideas?

I don't know about this error, but I got `mstream 4.3.0` working by doing:
```
cd
git clone https://github.com/IrosTheBeggar/mStream.git
cd mstream
git checkout relative-urls
npm install
sudo npm link
mstream -j /path/to/conf.json
```
any by adding a minimal /mstream/ location to the NginX configuration file:
```
location /mstream/   {
    proxy_pass http://127.0.0.1:3000/;
                                }
```
NginX access logs are fine and report the following:
```
192.168.1.101 - - [18/Sep/2019:22:58:47 +0200] "GET / HTTP/2.0" 200 [...]
192.168.1.101 - - [18/Sep/2019:22:58:48 +0200] "GET /mstream/ HTTP/2.0" 200 [...]
192.168.1.101 - - [18/Sep/2019:22:58:49 +0200] "GET /mstream/public/js/lib/vue.min.js HTTP/2.0" 503 [...] 
[...]
```

Using Arch Arm with `npm 6.11.3-1` and `nodejs 12.10.0-1`
@IrosTheBeggar  please merge the relative-urls branch into the master one

**EDIT:**
I've noticed the following:
- there is no `<noscript>` tag that alerts that javascript is disabled and thus the webapp cannot work
- something is requesting a non-existing .sh file in the /public/js/ folder, just before requesting "mstream.js":
```
151.xx.xx.xx - - [19/Sep/2019:19:19:30 +0200] "GET /mstream/public/js/mstream.sh HTTP/2.0" 503 753
151.xx.xx.xx - - [19/Sep/2019:19:19:30 +0200] "GET /mstream/public/js/mstream.js HTTP/2.0" 503 753
```
- the request for the Jura font points to a wrong path:
```
151.xx.xx.xx - - [19/Sep/2019:20:18:32 +0200] "GET /mstream/public/fonts/public/fonts/jura/z7NbdRfiaC4VXcRJUQZA3Jw.woff2 HTTP/2.0" 302 36 
```
- desktop browsers seems to work fine (Firefox, Chrome, TOR), but on iOS the webapp is inusable through NginX (it only works through 192.168.1.13:3000)

**Comment by Nixbligger** [comment]
@AlphaJack 
Thank you for sharing your experience. 
I added the trailing "/" to the subdir location and the proxy pass address in nginx and now the proxying seems to work - top! 

There is only one warning remaining on the Firefox console:

> Source-Map-Error: request failed with status 404
> Resourcen-Address: https://…/mstream/public/js/lib/popper.js
> Source-Map-Addresds popper.min.js.map


Now I still need to find a solution concerning the database scan not exceeding the 2250 entry mark...

@IrosTheBeggar: 
As I presume the database error might be related to some bad quality music files or tags - is there a way to check which of the files causes the hangs? Is there some kind of logging?

**Comment by Nixbligger** [comment]
Hello again,

as the initial issue was solved by the help of @AlphaJack, I'll try to mark this one as solved, and opened another issue #208 for the java heap memory database problem.

Thx and Cheers,
Nix

**Comment by maxim-kukushkin** [comment]
@Nixbligger, I'm not sure the original issue has been resolved. @AlphaJack has indicated that a branch worked, but not the main version of the code. For the main version of the code it's still there and is super annoying.
Any chance it can be re-opened and resolved?
